File: binding/src/libqpagesize.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qpagesize.html
pub const qpagesize = struct {

    /// New constructs a new QPageSize object.
    pub fn New() QtC.QPageSize {
        return qtc.QPageSize_new();
    }


    /// New2 constructs a new QPageSize object.
    pub fn New2(pageSizeId: i32) QtC.QPageSize {
        return qtc.QPageSize_new2(@intCast(pageSizeId));
    }


    /// New3 constructs a new QPageSize object.
    pub fn New3(pointSize: ?*anyopaque) QtC.QPageSize {
        return qtc.QPageSize_new3(@ptrCast(pointSize));
    }


    /// New4 constructs a new QPageSize object.
    pub fn New4(size: ?*anyopaque, units: i32) QtC.QPageSize {
        return qtc.QPageSize_new4(@ptrCast(size), @intCast(units));
    }


    /// New5 constructs a new QPageSize object.
    pub fn New5(other: ?*anyopaque) QtC.QPageSize {
        return qtc.QPageSize_new5(@ptrCast(other));
    }


    /// New6 constructs a new QPageSize object.
    pub fn New6(pointSize: ?*anyopaque, name: []const u8) QtC.QPageSize {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QPageSize_new6(@ptrCast(pointSize), name_str);
    }


    /// New7 constructs a new QPageSize object.
    pub fn New7(pointSize: ?*anyopaque, name: []const u8, matchPolicy: i32) QtC.QPageSize {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QPageSize_new7(@ptrCast(pointSize), name_str, @intCast(matchPolicy));
    }


    /// New8 constructs a new QPageSize object.
    pub fn New8(size: ?*anyopaque, units: i32, name: []const u8) QtC.QPageSize {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QPageSize_new8(@ptrCast(size), @intCast(units), name_str);
    }


    /// New9 constructs a new QPageSize object.
    pub fn New9(size: ?*anyopaque, units: i32, name: []const u8, matchPolicy: i32) QtC.QPageSize {
        const name_str = qtc.libqt_string{
    .len = name.len,
    .data = name.ptr,
};
return qtc.QPageSize_new9(@ptrCast(size), @intCast(units), name_str, @intCast(matchPolicy));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageSize_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QPageSize_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsEquivalentTo(self: ?*anyopaque, other: ?*anyopaque) bool {
        return qtc.QPageSize_IsEquivalentTo(@ptrCast(self), @ptrCast(other));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QPageSize_IsValid(@ptrCast(self));
    }

    pub fn Key(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPageSize_Key(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpagesize.Key: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Name(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPageSize_Name(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpagesize.Name: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Id(self: ?*anyopaque, ) i32 {
        return qtc.QPageSize_Id(@ptrCast(self));
    }

    pub fn WindowsId(self: ?*anyopaque, ) i32 {
        return qtc.QPageSize_WindowsId(@ptrCast(self));
    }

    pub fn DefinitionSize(self: ?*anyopaque, ) QtC.QSizeF {
        return qtc.QPageSize_DefinitionSize(@ptrCast(self));
    }

    pub fn DefinitionUnits(self: ?*anyopaque, ) i32 {
        return qtc.QPageSize_DefinitionUnits(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, units: i32) QtC.QSizeF {
        return qtc.QPageSize_Size(@ptrCast(self), @intCast(units));
    }

    pub fn SizePoints(self: ?*anyopaque, ) QtC.QSize {
        return qtc.QPageSize_SizePoints(@ptrCast(self));
    }

    pub fn SizePixels(self: ?*anyopaque, resolution: i32) QtC.QSize {
        return qtc.QPageSize_SizePixels(@ptrCast(self), resolution);
    }

    pub fn Rect(self: ?*anyopaque, units: i32) QtC.QRectF {
        return qtc.QPageSize_Rect(@ptrCast(self), @intCast(units));
    }

    pub fn RectPoints(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QPageSize_RectPoints(@ptrCast(self));
    }

    pub fn RectPixels(self: ?*anyopaque, resolution: i32) QtC.QRect {
        return qtc.QPageSize_RectPixels(@ptrCast(self), resolution);
    }

    pub fn Key2(pageSizeId: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPageSize_Key2(@intCast(pageSizeId));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpagesize.Key2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Name2(pageSizeId: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPageSize_Name2(@intCast(pageSizeId));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qpagesize.Name2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Id2(pointSize: ?*anyopaque) i32 {
        return qtc.QPageSize_Id2(@ptrCast(pointSize));
    }

    pub fn Id3(size: ?*anyopaque, units: i32) i32 {
        return qtc.QPageSize_Id3(@ptrCast(size), @intCast(units));
    }

    pub fn Id4(windowsId: i32) i32 {
        return qtc.QPageSize_Id4(windowsId);
    }

    pub fn WindowsId2(pageSizeId: i32) i32 {
        return qtc.QPageSize_WindowsId2(@intCast(pageSizeId));
    }

    pub fn DefinitionSize2(pageSizeId: i32) QtC.QSizeF {
        return qtc.QPageSize_DefinitionSize2(@intCast(pageSizeId));
    }

    pub fn DefinitionUnits2(pageSizeId: i32) i32 {
        return qtc.QPageSize_DefinitionUnits2(@intCast(pageSizeId));
    }

    pub fn Size2(pageSizeId: i32, units: i32) QtC.QSizeF {
        return qtc.QPageSize_Size2(@intCast(pageSizeId), @intCast(units));
    }

    pub fn SizePoints2(pageSizeId: i32) QtC.QSize {
        return qtc.QPageSize_SizePoints2(@intCast(pageSizeId));
    }

    pub fn SizePixels2(pageSizeId: i32, resolution: i32) QtC.QSize {
        return qtc.QPageSize_SizePixels2(@intCast(pageSizeId), resolution);
    }

    pub fn Id22(pointSize: ?*anyopaque, matchPolicy: i32) i32 {
        return qtc.QPageSize_Id22(@ptrCast(pointSize), @intCast(matchPolicy));
    }

    pub fn Id32(size: ?*anyopaque, units: i32, matchPolicy: i32) i32 {
        return qtc.QPageSize_Id32(@ptrCast(size), @intCast(units), @intCast(matchPolicy));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPageSize_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/pagesize.html#types
pub const enums = struct {
    pub const PageSizeId = enum {
        pub const Letter: i32 = 0;
        pub const Legal: i32 = 1;
        pub const Executive: i32 = 2;
        pub const A0: i32 = 3;
        pub const A1: i32 = 4;
        pub const A2: i32 = 5;
        pub const A3: i32 = 6;
        pub const A4: i32 = 7;
        pub const A5: i32 = 8;
        pub const A6: i32 = 9;
        pub const A7: i32 = 10;
        pub const A8: i32 = 11;
        pub const A9: i32 = 12;
        pub const A10: i32 = 13;
        pub const B0: i32 = 14;
        pub const B1: i32 = 15;
        pub const B2: i32 = 16;
        pub const B3: i32 = 17;
        pub const B4: i32 = 18;
        pub const B5: i32 = 19;
        pub const B6: i32 = 20;
        pub const B7: i32 = 21;
        pub const B8: i32 = 22;
        pub const B9: i32 = 23;
        pub const B10: i32 = 24;
        pub const C5E: i32 = 25;
        pub const Comm10E: i32 = 26;
        pub const DLE: i32 = 27;
        pub const Folio: i32 = 28;
        pub const Ledger: i32 = 29;
        pub const Tabloid: i32 = 30;
        pub const Custom: i32 = 31;
        pub const A3Extra: i32 = 32;
        pub const A4Extra: i32 = 33;
        pub const A4Plus: i32 = 34;
        pub const A4Small: i32 = 35;
        pub const A5Extra: i32 = 36;
        pub const B5Extra: i32 = 37;
        pub const JisB0: i32 = 38;
        pub const JisB1: i32 = 39;
        pub const JisB2: i32 = 40;
        pub const JisB3: i32 = 41;
        pub const JisB4: i32 = 42;
        pub const JisB5: i32 = 43;
        pub const JisB6: i32 = 44;
        pub const JisB7: i32 = 45;
        pub const JisB8: i32 = 46;
        pub const JisB9: i32 = 47;
        pub const JisB10: i32 = 48;
        pub const AnsiC: i32 = 49;
        pub const AnsiD: i32 = 50;
        pub const AnsiE: i32 = 51;
        pub const LegalExtra: i32 = 52;
        pub const LetterExtra: i32 = 53;
        pub const LetterPlus: i32 = 54;
        pub const LetterSmall: i32 = 55;
        pub const TabloidExtra: i32 = 56;
        pub const ArchA: i32 = 57;
        pub const ArchB: i32 = 58;
        pub const ArchC: i32 = 59;
        pub const ArchD: i32 = 60;
        pub const ArchE: i32 = 61;
        pub const Imperial7x9: i32 = 62;
        pub const Imperial8x10: i32 = 63;
        pub const Imperial9x11: i32 = 64;
        pub const Imperial9x12: i32 = 65;
        pub const Imperial10x11: i32 = 66;
        pub const Imperial10x13: i32 = 67;
        pub const Imperial10x14: i32 = 68;
        pub const Imperial12x11: i32 = 69;
        pub const Imperial15x11: i32 = 70;
        pub const ExecutiveStandard: i32 = 71;
        pub const Note: i32 = 72;
        pub const Quarto: i32 = 73;
        pub const Statement: i32 = 74;
        pub const SuperA: i32 = 75;
        pub const SuperB: i32 = 76;
        pub const Postcard: i32 = 77;
        pub const DoublePostcard: i32 = 78;
        pub const Prc16K: i32 = 79;
        pub const Prc32K: i32 = 80;
        pub const Prc32KBig: i32 = 81;
        pub const FanFoldUS: i32 = 82;
        pub const FanFoldGerman: i32 = 83;
        pub const FanFoldGermanLegal: i32 = 84;
        pub const EnvelopeB4: i32 = 85;
        pub const EnvelopeB5: i32 = 86;
        pub const EnvelopeB6: i32 = 87;
        pub const EnvelopeC0: i32 = 88;
        pub const EnvelopeC1: i32 = 89;
        pub const EnvelopeC2: i32 = 90;
        pub const EnvelopeC3: i32 = 91;
        pub const EnvelopeC4: i32 = 92;
        pub const EnvelopeC6: i32 = 93;
        pub const EnvelopeC65: i32 = 94;
        pub const EnvelopeC7: i32 = 95;
        pub const Envelope9: i32 = 96;
        pub const Envelope11: i32 = 97;
        pub const Envelope12: i32 = 98;
        pub const Envelope14: i32 = 99;
        pub const EnvelopeMonarch: i32 = 100;
        pub const EnvelopePersonal: i32 = 101;
        pub const EnvelopeChou3: i32 = 102;
        pub const EnvelopeChou4: i32 = 103;
        pub const EnvelopeInvite: i32 = 104;
        pub const EnvelopeItalian: i32 = 105;
        pub const EnvelopeKaku2: i32 = 106;
        pub const EnvelopeKaku3: i32 = 107;
        pub const EnvelopePrc1: i32 = 108;
        pub const EnvelopePrc2: i32 = 109;
        pub const EnvelopePrc3: i32 = 110;
        pub const EnvelopePrc4: i32 = 111;
        pub const EnvelopePrc5: i32 = 112;
        pub const EnvelopePrc6: i32 = 113;
        pub const EnvelopePrc7: i32 = 114;
        pub const EnvelopePrc8: i32 = 115;
        pub const EnvelopePrc9: i32 = 116;
        pub const EnvelopePrc10: i32 = 117;
        pub const EnvelopeYou4: i32 = 118;
        pub const LastPageSize: i32 = 118;
        pub const AnsiA: i32 = 0;
        pub const AnsiB: i32 = 29;
        pub const EnvelopeC5: i32 = 25;
        pub const EnvelopeDL: i32 = 27;
        pub const Envelope10: i32 = 26;
    };

    pub const Unit = enum {
        pub const Millimeter: i32 = 0;
        pub const Point: i32 = 1;
        pub const Inch: i32 = 2;
        pub const Pica: i32 = 3;
        pub const Didot: i32 = 4;
        pub const Cicero: i32 = 5;
    };

    pub const SizeMatchPolicy = enum {
        pub const FuzzyMatch: i32 = 0;
        pub const FuzzyOrientationMatch: i32 = 1;
        pub const ExactMatch: i32 = 2;
    };

};

File: binding/src/libqpagesize.cpp
#include <QPageSize>
#include <QRect>
#include <QRectF>
#include <QSize>
#include <QSizeF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qpagesize.h>
#include "libqpagesize.h"
#include "libqpagesize.hxx"

QPageSize* QPageSize_new() {
	 return new QPageSize();
}

QPageSize* QPageSize_new2(int pageSizeId) {
	 return new QPageSize(static_cast<QPageSize::PageSizeId>(pageSizeId));
}

QPageSize* QPageSize_new3(const QSize* pointSize) {
	 return new QPageSize(*pointSize);
}

QPageSize* QPageSize_new4(const QSizeF* size, int units) {
	 return new QPageSize(*size, static_cast<QPageSize::Unit>(units));
}

QPageSize* QPageSize_new5(const QPageSize* other) {
	 return new QPageSize(*other);
}

QPageSize* QPageSize_new6(const QSize* pointSize, const libqt_string name) {
	 return new QPageSize(*pointSize, QString::fromUtf8(name.data, name.len));
}

QPageSize* QPageSize_new7(const QSize* pointSize, const libqt_string name, int matchPolicy) {
	 return new QPageSize(*pointSize, QString::fromUtf8(name.data, name.len), static_cast<QPageSize::SizeMatchPolicy>(matchPolicy));
}

QPageSize* QPageSize_new8(const QSizeF* size, int units, const libqt_string name) {
	 return new QPageSize(*size, static_cast<QPageSize::Unit>(units), QString::fromUtf8(name.data, name.len));
}

QPageSize* QPageSize_new9(const QSizeF* size, int units, const libqt_string name, int matchPolicy) {
	 return new QPageSize(*size, static_cast<QPageSize::Unit>(units), QString::fromUtf8(name.data, name.len), static_cast<QPageSize::SizeMatchPolicy>(matchPolicy));
}

void QPageSize_OperatorAssign(QPageSize* self, const QPageSize* other) {
	self->operator=(*other);
}

void QPageSize_Swap(QPageSize* self, QPageSize* other) {
	self->swap(*other);
}

bool QPageSize_IsEquivalentTo(const QPageSize* self, const QPageSize* other) {
	return self->isEquivalentTo(*other);
}

bool QPageSize_IsValid(const QPageSize* self) {
	return self->isValid();
}

libqt_string QPageSize_Key(const QPageSize* self) {
	QString _ret = self->key();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPageSize_Name(const QPageSize* self) {
	QString _ret = self->name();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QPageSize_Id(const QPageSize* self) {
	return self->id();
}

int QPageSize_WindowsId(const QPageSize* self) {
	return self->windowsId();
}

QSizeF* QPageSize_DefinitionSize(const QPageSize* self) {
	return new QSizeF(self->definitionSize());
}

int QPageSize_DefinitionUnits(const QPageSize* self) {
	return self->definitionUnits();
}

QSizeF* QPageSize_Size(const QPageSize* self, int units) {
	return new QSizeF(self->size(static_cast<QPageSize::Unit>(units)));
}

QSize* QPageSize_SizePoints(const QPageSize* self) {
	return new QSize(self->sizePoints());
}

QSize* QPageSize_SizePixels(const QPageSize* self, int resolution) {
	return new QSize(self->sizePixels(resolution));
}

QRectF* QPageSize_Rect(const QPageSize* self, int units) {
	return new QRectF(self->rect(static_cast<QPageSize::Unit>(units)));
}

QRect* QPageSize_RectPoints(const QPageSize* self) {
	return new QRect(self->rectPoints());
}

QRect* QPageSize_RectPixels(const QPageSize* self, int resolution) {
	return new QRect(self->rectPixels(resolution));
}

libqt_string QPageSize_Key2(int pageSizeId) {
	QString _ret = QPageSize::key(static_cast<QPageSize::PageSizeId>(pageSizeId));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPageSize_Name2(int pageSizeId) {
	QString _ret = QPageSize::name(static_cast<QPageSize::PageSizeId>(pageSizeId));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QPageSize_Id2(const QSize* pointSize) {
	return QPageSize::id(*pointSize);
}

int QPageSize_Id3(const QSizeF* size, int units) {
	return QPageSize::id(*size, static_cast<QPageSize::Unit>(units));
}

int QPageSize_Id4(int windowsId) {
	return QPageSize::id(windowsId);
}

int QPageSize_WindowsId2(int pageSizeId) {
	return QPageSize::windowsId(static_cast<QPageSize::PageSizeId>(pageSizeId));
}

QSizeF* QPageSize_DefinitionSize2(int pageSizeId) {
	return new QSizeF(QPageSize::definitionSize(static_cast<QPageSize::PageSizeId>(pageSizeId)));
}

int QPageSize_DefinitionUnits2(int pageSizeId) {
	return QPageSize::definitionUnits(static_cast<QPageSize::PageSizeId>(pageSizeId));
}

QSizeF* QPageSize_Size2(int pageSizeId, int units) {
	return new QSizeF(QPageSize::size(static_cast<QPageSize::PageSizeId>(pageSizeId), static_cast<QPageSize::Unit>(units)));
}

QSize* QPageSize_SizePoints2(int pageSizeId) {
	return new QSize(QPageSize::sizePoints(static_cast<QPageSize::PageSizeId>(pageSizeId)));
}

QSize* QPageSize_SizePixels2(int pageSizeId, int resolution) {
	return new QSize(QPageSize::sizePixels(static_cast<QPageSize::PageSizeId>(pageSizeId), resolution));
}

int QPageSize_Id22(const QSize* pointSize, int matchPolicy) {
	return QPageSize::id(*pointSize, static_cast<QPageSize::SizeMatchPolicy>(matchPolicy));
}

int QPageSize_Id32(const QSizeF* size, int units, int matchPolicy) {
	return QPageSize::id(*size, static_cast<QPageSize::Unit>(units), static_cast<QPageSize::SizeMatchPolicy>(matchPolicy));
}

void QPageSize_Delete(QPageSize* self) {
    delete self;
}

